gtk/tr-prefs.c

```diff
@@ -197,6 +197,45 @@ static GtkWidget* new_entry(tr_quark const key, gpointer core)
     return w;
 }
 
+static void text_buffer_changed_cb(GtkTextBuffer* buffer, gpointer core)
+{
+    tr_quark const key = GPOINTER_TO_INT(g_object_get_data(G_OBJECT(buffer), PREF_KEY));
+    GtkTextIter start, end;
+    gtk_text_buffer_get_start_iter(buffer, &start);
+    gtk_text_buffer_get_end_iter(buffer, &end);
+    char const* value = gtk_text_buffer_get_text(buffer, &start, &end, FALSE);
+
+    gtr_core_set_pref(TR_CORE(core), key, value);
+}
+
+static GtkWidget* new_text_view(tr_quark const key, gpointer core)
+{
+    GtkWidget* scroll = gtk_scrolled_window_new(NULL, NULL);
+    GtkWidget* w = gtk_text_view_new();
+    GtkTextBuffer* buffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(w));
+    char const* value = gtr_pref_string_get(key);
+
+    if (value)
+    {
+        gtk_text_buffer_set_text(buffer, value, -1);
+    }
+
+    /* set up the scrolled window and put the text view in it */
+    gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(scroll),
+                                   GTK_POLICY_AUTOMATIC,
+                                   GTK_POLICY_AUTOMATIC);
+    gtk_scrolled_window_set_shadow_type(GTK_SCROLLED_WINDOW(scroll),
+                                        GTK_SHADOW_IN);
+    gtk_container_add(GTK_CONTAINER(scroll), w);
+    gtk_widget_set_size_request(scroll, -1, 200);
+
+    /* key and signal */
+    g_object_set_data(G_OBJECT(buffer), PREF_KEY, GINT_TO_POINTER(key));
+    g_signal_connect(buffer, "changed", G_CALLBACK(text_buffer_changed_cb), core);
```

Comment:
I _think_ this is OK but my GTK+ is slightly rusty; can you confirm this event isn't fired off for every keystroke?

Reply:
https://developer.gnome.org/gtk3/stable/GtkEditable.html#GtkEditable-changed :

> The ::changed signal is emitted at the end of a single user-visible operation on the contents of the GtkEditable.
> 
> E.g., a paste operation that replaces the contents of the selection will cause only one signal emission (even though it is implemented by first deleting the selection, then inserting the new content, and may cause multiple ::notify::text signals to be emitted).
>